tests/test_builders.py

```diff
@@ -0,0 +1,99 @@
+import os
+
+import pytest
+import sphinx
+from sphinx import addnodes
+from sphinx.builders.html import SingleFileHTMLBuilder, DirectoryHTMLBuilder
+
+from .util import build_all
+
+
+def test_basic():
+    for (app, status, warning) in build_all('test-basic'):
+        assert app.env.get_doctree('index').traverse(addnodes.toctree)
+        content = open(os.path.join(app.outdir, 'index.html')).read()
+
+        if isinstance(app.builder, DirectoryHTMLBuilder):
+            search = (
+                '<div class="toctree-wrapper compound">\n'
+                '<ul>\n'
+                '<li class="toctree-l1">'
+                '<a class="reference internal" href="foo/">foo</a>'
+                '<ul>\n'
+                '<li class="toctree-l2">'
+                '<a class="reference internal" href="bar/">bar</a></li>\n'
+                '</ul>\n'
+                '</li>\n'
+                '</ul>\n'
+                '</div>'
+            )
+            assert search in content
+        elif isinstance(app.builder, SingleFileHTMLBuilder):
+            search = (
+                '<div class="local-toc"><ul>\n'
+                '<li class="toctree-l1">'
+                '<a class="reference internal" href="index.html#document-foo">foo</a>'
+                '<ul>\n'
+                '<li class="toctree-l2">'
+                '<a class="reference internal" href="index.html#document-bar">bar</a>'
+                '</li>\n'
+                '</ul>'
+            )
```

Comment:
It shouldn't be all that brittle, but I do think it's important to test the full thing. Otherwise, we won't be testing if the toctrees have anything in them, which is the more important part.